src/chain_actions.cc

```diff
@@ -336,6 +337,149 @@ namespace dd
+          {
+            APIData bbox = ad_cls.at(j).getobj("bbox");
+            std::string cat = ad_cls.at(j).get("cat").get<std::string>();
+            if (bbox.empty())
+              throw ActionBadParamException(
+                  "draw action cannot find bbox object for uri " + uri);
+
+            double xmin = bbox.get("xmin").get<double>() / orig_cols * im_cols;
+            double ymin = bbox.get("ymin").get<double>() / orig_rows * im_rows;
+            double xmax = bbox.get("xmax").get<double>() / orig_cols * im_cols;
+            double ymax = bbox.get("ymax").get<double>() / orig_rows * im_rows;
+
+            // draw bbox
+            cv::Point pt1{ int(xmin), int(ymin) };
+            cv::Point pt2{ int(xmax), int(ymax) };
+            size_t cls_hash = std::hash<std::string>{}(cat);
+            cv::Scalar color = bbox_palette[cls_hash % bbox_palette_size];
+            cv::rectangle(rimg, pt1, pt2, cv::Scalar(255, 255, 255),
+                          ref_thickness + 2);
+            cv::rectangle(rimg, pt1, pt2, color, ref_thickness);
+
+            // draw class & confidences
+            std::string label;
+            if (_params->write_cat)
+              label = cat;
+            if (_params->write_cat && _params->write_prob)
+              label += " - ";
+            if (_params->write_prob)
+              label += std::to_string(ad_cls.at(j).get("prob").get<double>());
+
+            // font size relatively to base opencv font size
+            float font_size = 2;
+            int x_txt = static_cast<int>(xmin + 5);
+            if (x_txt > im_cols - 15)
+              x_txt = im_cols - 15;
+            int y_txt = std::min(im_rows - 20,
+                                 static_cast<int>(ymax + 2 + font_size * 12));
+
+            cv::putText(rimg, label, cv::Point(x_txt, y_txt),
+                        cv::FONT_HERSHEY_PLAIN, font_size,
+                        cv::Scalar(255, 255, 255), ref_thickness + 2);
+            cv::putText(rimg, label, cv::Point(x_txt, y_txt),
+                        cv::FONT_HERSHEY_PLAIN, font_size, color,
+                        ref_thickness);
+          }
+
+        rimgs.push_back(rimg);
+
+        // save image if requested
+        if (save_img)
+          {
+            std::string puri = dd_utils::split(uri, '/').back();
+            std::string rimg_path = save_path + "bbox_" + puri + ".png";
+            this->_chain_logger->info("draw_bbox: Saved image to path {}",
+                                      rimg_path);
+            cv::imwrite(rimg_path, rimg);
+          }
+
+        if (_params->output_images)
+          {
+            auto action_pred = DTO::Prediction::createShared();
+
+            // image as float array, same format as GAN
+            // n_chan * width * height
+            std::vector<double> img_vec;
+            img_vec.reserve(rimg.channels() * rimg.rows * rimg.cols);
+
+            for (int c = 0; c < rimg.channels(); ++c)
```

Comment:
Is there any vectorized version of this in OpenCV ?

Reply:
I'm swapping channels to match the output format of torch GAN. I think at this point it's simpler to keep it like that.

Reply:
Shouldn't we swap only when `rgb` is set to true ?

Reply:
I removed the transformation, now the raw image is returned